## 1. PDGA APPROVED Discs
load pdga approved discs from csv file "data/pdga_approved.csv"

In [59]:
import pandas as pd
import numpy as np

In [60]:
dimensions_df = pd.read_csv("data/pdga_approved.csv",skiprows=1)
dimensions_df.head()

Manufacturer / Distributor        Disc Model  Max Weight (gr)  \
0            Full Turn Discs             Guide            180.9   
1          Wing It Disc Golf            Dream             176.8   
2              Dynamic Discs   Sockibomb Felon            176.0   
3              Dynamic Discs  Supreme Fugitive            180.1   
4              STM Disc Golf      Cloud Ripper            175.1   

   Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0           21.8          1.8             1.2                      19.0   
1           21.3          1.4             1.1                      16.5   
2           21.2          1.6             1.1                      17.1   
3           21.7          1.9             1.3                      18.7   
4           21.1          2.1             1.4                      18.8   

   Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                 1.4                             5.5              55.50   
1                 2.4                             5.2              27.00   
2                 2.0                             5.2              31.50   
3                 1.5                             6.0              39.00   
4                 1.2                             6.6              46.25   

   Flexibility (kg) Class  Max Weight Vint (gr)  Last Year Production  \
0              6.36   NaN                   NaN                   NaN   
1              9.32   NaN                   NaN                   NaN   
2             11.93   NaN                   NaN                   NaN   
3             11.70   NaN                   NaN                   NaN   
4              8.75   NaN                   NaN                   NaN   

  Certification Number Approved Date  
0                23-20  Jan 23, 2023  
1                23-21  Jan 23, 2023  
2                23-18  Jan 23, 2023  
3                23-19  Jan 23, 2023  
4                23-12  Jan 16, 2023

In [61]:
dimensions_df.describe()

Max Weight (gr)  Diameter (cm)  Height (cm)  Rim Depth (cm)  \
count      1674.000000    1674.000000  1674.000000     1674.000000   
mean        177.443489      21.421505     1.876822        1.283751   
std           4.525150       0.810299     0.331510        0.185709   
min         174.300000      21.000000     1.300000        1.100000   
25%         175.100000      21.100000     1.700000        1.100000   
50%         176.000000      21.200000     1.800000        1.200000   
75%         177.600000      21.400000     2.000000        1.400000   
max         200.000000      27.600000     3.800000        2.300000   

       Inside Rim Diameter (cm)  Rim Thickness (cm)  \
count               1674.000000         1674.000000   
mean                  18.231840            1.591637   
std                    1.529419            0.520714   
min                   14.500000            0.300000   
25%                   17.200000            1.100000   
50%                   18.300000            1.600000   
75%                   19.100000            2.000000   
max                   26.300000            4.100000   

       Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
count                     1674.000000        1674.000000       1672.000000   
mean                         5.993190          42.562939          8.724683   
std                          0.776055          17.419985          2.383915   
min                          5.000000          26.000000          0.560000   
25%                          5.200000          29.062500          7.260000   
50%                          5.700000          35.500000          8.960000   
75%                          6.500000          52.500000         10.660000   
max                         11.800000         114.250000         12.270000   

       Max Weight Vint (gr)  Last Year Production  
count             58.000000                   0.0  
mean             162.236207                   NaN  
std               15.360906                   NaN  
min              140.700000                   NaN  
25%              150.800000                   NaN  
50%              159.450000                   NaN  
75%              178.725000                   NaN  
max              184.300000                   NaN

In [62]:
dimensions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1674 entries, 0 to 1673
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Manufacturer / Distributor      1674 non-null   object 
 1   Disc Model                      1674 non-null   object 
 2   Max Weight (gr)                 1674 non-null   float64
 3   Diameter (cm)                   1674 non-null   float64
 4   Height (cm)                     1674 non-null   float64
 5   Rim Depth (cm)                  1674 non-null   float64
 6   Inside Rim Diameter (cm)        1674 non-null   float64
 7   Rim Thickness (cm)              1674 non-null   float64
 8   Rim Depth / Diameter Ratio (%)  1674 non-null   float64
 9   Rim Configuration               1674 non-null   float64
 10  Flexibility (kg)                1672 non-null   float64
 11  Class                           63 non-null     object 
 12  Max Weight Vint (gr)            58

In [63]:
##drop columns
dimensions_df = dimensions_df.drop(["Class", "Max Weight Vint (gr)", "Last Year Production"], axis=1)
dimensions_df.head()

Manufacturer / Distributor        Disc Model  Max Weight (gr)  \
0            Full Turn Discs             Guide            180.9   
1          Wing It Disc Golf            Dream             176.8   
2              Dynamic Discs   Sockibomb Felon            176.0   
3              Dynamic Discs  Supreme Fugitive            180.1   
4              STM Disc Golf      Cloud Ripper            175.1   

   Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0           21.8          1.8             1.2                      19.0   
1           21.3          1.4             1.1                      16.5   
2           21.2          1.6             1.1                      17.1   
3           21.7          1.9             1.3                      18.7   
4           21.1          2.1             1.4                      18.8   

   Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                 1.4                             5.5              55.50   
1                 2.4                             5.2              27.00   
2                 2.0                             5.2              31.50   
3                 1.5                             6.0              39.00   
4                 1.2                             6.6              46.25   

   Flexibility (kg) Certification Number Approved Date  
0              6.36                23-20  Jan 23, 2023  
1              9.32                23-21  Jan 23, 2023  
2             11.93                23-18  Jan 23, 2023  
3             11.70                23-19  Jan 23, 2023  
4              8.75                23-12  Jan 16, 2023

## Scraping using Beautiful Soup


In [64]:
from bs4 import BeautifulSoup as soup
import re
import requests

In [65]:
#using a saved version of the html page to not disrupt during testing
my_page = "data/all_disc_test.html"
souped = soup(open(my_page),"html")

In [66]:
#get the classes of each table
print('Classes of each table:')
for table in souped.find_all('table'):
    print(table.get('id'))

Classes of each table:
table_1


### Get the Table Headers

In [67]:
table = souped.find('table',{"id":"table_1"})
header = table.findAll('th')
headers=[]
#don't include the options row
for h in header[:9]:
    headers.append(h.text)
headers

['Brand',
 'Mold',
 'Type',
 'Speed',
 'Glide',
 'Turn',
 'Fade',
 'In Prod.',
 'Date Approved']

### Get the Data!
Now we need to start at spot 2 to get all of the disc data this will use td and create a list of lists for each 9 items.  There are 9 columns in the table.

In [68]:
#list of all discs
disc_data=[]
#data with all the discs 
disc_text = table.findAll('td')
#empty temp list and counter starting at 0
counter = 0
temp_list=[]

#loop through each 9 cells and create a row for each disc.
for d in disc_text:
    d_text = d.text
    if counter <=8:
        temp_list.append(d_text)
        counter+=1
    else:
        counter=1
        disc_data.append(temp_list)
        temp_list=[]
        temp_list.append(d_text)



In [69]:
#show first 3 discs
disc_data[:3]
    

[['ABC Discs',
  'Bee Line',
  'Distance',
  '14.0',
  '5.0',
  '-3.0',
  '2.0',
  'Yes',
  '09/12/2012'],
 ['ABC Discs',
  'Flying Squirrel',
  'Midrange',
  '6.0',
  '5.0',
  '-3.0',
  '2.0',
  'Yes',
  '05/29/2010'],
 ['ABC Discs',
  'Mission',
  'Midrange',
  '4.0',
  '5.0',
  '-1.0',
  '2.0',
  'Yes',
  '07/27/2010']]

## Create Dataframe from the two lists

Using our disc_data with the headers list as column titles to create a dataframe.

In [70]:
flight_number_df = pd.DataFrame(disc_data,columns = headers)
flight_number_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved  
0    09/12/2012  
1    05/29/2010  
2    07/27/2010  
3    04/09/2010  
4    05/29/2010

In [71]:
flight_number_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 0 to 742
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Brand          743 non-null    object
 1   Mold           743 non-null    object
 2   Type           743 non-null    object
 3   Speed          743 non-null    object
 4   Glide          743 non-null    object
 5   Turn           743 non-null    object
 6   Fade           743 non-null    object
 7   In Prod.       743 non-null    object
 8   Date Approved  743 non-null    object
dtypes: object(9)
memory usage: 52.4+ KB


## Merging the Data Sets 
This will be a challenge because the mold names will not match exactly and are inconsistent.  We will do our best.

### Fix The Date Approved Columns
Both date columns are currently objects that could be switched into datetime64.  They are formatted differently so we will switch the data types and use the default format for both.

This process will help us match the discs that we can't match by name.

In [72]:
##dimensions_df - Approved Date to datetime64
dimensions_df['Date Approved'] = pd.to_datetime(dimensions_df['Approved Date'])
dimensions_df.head()

Manufacturer / Distributor        Disc Model  Max Weight (gr)  \
0            Full Turn Discs             Guide            180.9   
1          Wing It Disc Golf            Dream             176.8   
2              Dynamic Discs   Sockibomb Felon            176.0   
3              Dynamic Discs  Supreme Fugitive            180.1   
4              STM Disc Golf      Cloud Ripper            175.1   

   Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0           21.8          1.8             1.2                      19.0   
1           21.3          1.4             1.1                      16.5   
2           21.2          1.6             1.1                      17.1   
3           21.7          1.9             1.3                      18.7   
4           21.1          2.1             1.4                      18.8   

   Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                 1.4                             5.5              55.50   
1                 2.4                             5.2              27.00   
2                 2.0                             5.2              31.50   
3                 1.5                             6.0              39.00   
4                 1.2                             6.6              46.25   

   Flexibility (kg) Certification Number Approved Date Date Approved  
0              6.36                23-20  Jan 23, 2023    2023-01-23  
1              9.32                23-21  Jan 23, 2023    2023-01-23  
2             11.93                23-18  Jan 23, 2023    2023-01-23  
3             11.70                23-19  Jan 23, 2023    2023-01-23  
4              8.75                23-12  Jan 16, 2023    2023-01-16

In [73]:
##flight_number_df Date Approved change
flight_number_df['Date Approved'] = pd.to_datetime(flight_number_df['Date Approved'])
flight_number_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved  
0    2012-09-12  
1    2010-05-29  
2    2010-07-27  
3    2010-04-09  
4    2010-05-29

### Looking at Flight Number Data

In [74]:
#discmania brand only view
flight_number_df[flight_number_df['Brand']=='Discmania']

Brand                   Mold       Type Speed Glide  Turn Fade  \
92   Discmania             CD - Craze   Distance  10.0   5.0  -1.0  2.0   
93   Discmania                    CD2    Fairway   9.0   5.0  -1.0  2.0   
94   Discmania                    CD3  Distance   11.0   5.0  -1.0  2.0   
95   Discmania          DD - Hysteria   Distance  11.0   5.0   0.0  2.0   
96   Discmania             DD2 Frenzy   Distance  13.0   5.0  -2.0  2.0   
97   Discmania                    DDx   Distance  12.0   6.0  -1.0  2.0   
98   Discmania            FD - Jackal    Fairway   7.0   6.0  -1.0  1.0   
99   Discmania                    FD2    Fairway   7.0   4.0   0.0  2.0   
100  Discmania         FD2 (retooled)    Fairway   7.0   4.0   0.0  2.0   
101  Discmania                    FD3    Fairway   9.0   4.0   0.0  3.0   
102  Discmania                     MD   Midrange   4.0   5.0   0.0  0.0   
103  Discmania  MD1 - Midrange Driver   Midrange   4.0   4.0   0.0  2.0   
104  Discmania            MD2 - Fiend   Midrange   4.0   5.0   0.0  2.0   
105  Discmania  MD3 - Midrange Driver   Midrange   5.0   5.0   0.0  2.0   
106  Discmania                    MD4   Midrange   5.0   4.0   0.0  3.0   
107  Discmania                    MD5   Midrange   5.0   4.0   0.0  2.0   
108  Discmania            P1 - Maniac     Putter   2.0   3.0   0.0  0.0   
109  Discmania                    P1x     Putter   2.0   3.0   0.0  0.0   
110  Discmania            P2 - Psycho     Putter   2.0   3.0   0.0  1.0   
111  Discmania   P3 - Putt & Approach     Putter   3.0   2.0   0.0  3.0   
112  Discmania                    P3x     Putter   3.0   2.0   0.0  3.0   
113  Discmania             PD - Freak   Distance  10.0   4.0   0.0  3.0   
114  Discmania                    PDx   Distance  11.0   4.0   0.0  3.0   
115  Discmania            PD2 - Chaos   Distance  12.0   4.0   0.0  4.0   
116  Discmania              TD - Rush   Distance  10.0   5.0  -2.0  1.0   
117  Discmania            TD2 - Fever   Distance  10.0   5.0  -4.0  1.0   
118  Discmania                    TDx    Fairway   9.0   5.0  -3.0  1.0   
712  Discmania            Sea Serpent   Distance  12.0   5.0  -2.0  2.0   
713  Discmania             Fox Spirit   Midrange   6.0   4.0   0.0  2.0   
714  Discmania               Instinct    Fairway   8.0   5.0   0.0  2.0   
715  Discmania               Sun Crow    Fairway   8.0   5.0  -4.0  1.0   
716  Discmania              Spring Ox   Midrange   4.0   3.0   0.0  2.0   
717  Discmania          Guardian Lion     Putter   1.0   3.0   0.0  1.0   
718  Discmania                    DD3   Distance  12.0   5.0  -1.0  3.0   
724  Discmania                 Method   Midrange   5.0   5.0   0.0  3.0   
728  Discmania                 Enigma   Distance  13.0   5.0  -1.0  3.0   
731  Discmania                   Link     Putter   2.0   3.0   0.0  1.0   

    In Prod. Date Approved  
92       Yes    2007-03-08  
93       Yes    2014-09-05  
94       Yes    2017-03-08  
95       Yes    2008-08-26  
96       Yes    2010-08-18  
97       Yes    2016-02-25  
98       Yes    2011-11-13  
99       Yes    2014-03-24  
100      Yes    2018-03-29  
101      Yes    2015-06-26  
102      Yes    2015-04-29  
103      Yes    2006-11-14  
104      Yes    2010-04-04  
105      Yes    2013-03-29  
106      Yes    2017-04-21  
107      Yes    2018-03-29  
108      Yes    2007-03-08  
109      Yes    2014-10-21  
110      Yes    2007-03-08  
111      Yes    2013-02-13  
112      Yes    2016-06-26  
113      Yes    2008-07-02  
114      Yes    2015-12-09  
115      Yes    2010-12-03  
116      Yes    2010-04-04  
117      Yes    2012-09-12  
118      Yes    2017-04-21  
712      Yes    2019-04-03  
713      Yes    2019-04-03  
714      Yes    2019-02-27  
715      Yes    2019-04-03  
716      Yes    2019-04-03  
717      Yes    2019-04-03  
718      Yes    2019-04-17  
724      Yes    2019-05-01  
728      Yes    2019-05-23  
731      Yes    2019-05-23

### Looking at Dimension Data
we will attempt to match as many of the 743 discs as possible

In [75]:
#drop the Approved Date Column
dimensions_df = dimensions_df.drop('Approved Date', axis=1)
dimensions_df[dimensions_df['Manufacturer / Distributor']=='Discmania']

Manufacturer / Distributor                    Disc Model  \
36                    Discmania                           FD1   
41                    Discmania                        Tailor   
42                    Discmania                           CD1   
43                    Discmania                           DD1   
71                    Discmania                       Mermaid   
...                         ...                           ...   
1260                  Discmania      PD (Freak, Power Driver)   
1307                  Discmania    CD (Craze, Control Driver)   
1308                  Discmania  P1 (Maniac, Putt & Approach)   
1311                  Discmania       P2 (Psycho, Pro Putter)   
1329                  Discmania         MD1 (Midrange Driver)   

      Max Weight (gr)  Diameter (cm)  Height (cm)  Rim Depth (cm)  \
36              176.0           21.2          1.7             1.1   
41              180.1           21.7          1.8             1.3   
42              176.0           21.2          1.7             1.1   
43              176.0           21.2          1.5             1.1   
71              176.8           21.3          1.8             1.2   
...               ...            ...          ...             ...   
1260            175.1           21.1          1.7             1.2   
1307            176.0           21.2          1.3             1.1   
1308            176.0           21.2          2.0             1.5   
1311            176.0           21.2          2.1             1.5   
1329            180.1           21.7          2.2             1.3   

      Inside Rim Diameter (cm)  Rim Thickness (cm)  \
36                        17.6                 1.8   
41                        19.1                 1.3   
42                        17.3                 1.9   
43                        16.8                 2.2   
71                        17.5                 1.9   
...                        ...                 ...   
1260                      17.3                 1.9   
1307                      17.3                 1.9   
1308                      19.3                 0.9   
1311                      19.2                 1.0   
1329                      19.2                 1.3   

      Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
36                               5.2              36.25              8.98   
41                               6.0              45.25              8.86   
42                               5.2              31.25              8.52   
43                               5.2              30.75              8.30   
71                               5.6              41.75              8.41   
...                              ...                ...               ...   
1260                             5.7              28.25             11.11   
1307                             5.2              28.50             10.43   
1308                             7.1              56.00              6.01   
1311                             7.1              58.25              7.03   
1329                             6.0              40.50              8.50   

     Certification Number Date Approved  
36                 22-211    2022-11-28  
41                 22-212    2022-11-28  
42                 22-209    2022-11-28  
43                 22-210    2022-11-28  
71                 22-178    2022-10-17  
...                   ...           ...  
1260                08-22    2008-07-02  
1307                07-14    2007-03-08  
1308                07-12    2007-03-08  
1311                07-13    2007-03-08  
1329                06-47    2006-11-14  

[69 rows x 13 columns]

## Names of molds do not match 
In order to be efficient we will only work with the manufactures that are in the smalller flight number dataframe and change the names of columns to match.

In [76]:
##change column names

dimensions_df.rename(
    columns={"Manufacturer / Distributor": "Brand", "Disc Model": "Mold"},
    inplace=True,
)

dimensions_df.head()

Brand              Mold  Max Weight (gr)  Diameter (cm)  \
0    Full Turn Discs             Guide            180.9           21.8   
1  Wing It Disc Golf            Dream             176.8           21.3   
2      Dynamic Discs   Sockibomb Felon            176.0           21.2   
3      Dynamic Discs  Supreme Fugitive            180.1           21.7   
4      STM Disc Golf      Cloud Ripper            175.1           21.1   

   Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  Rim Thickness (cm)  \
0          1.8             1.2                      19.0                 1.4   
1          1.4             1.1                      16.5                 2.4   
2          1.6             1.1                      17.1                 2.0   
3          1.9             1.3                      18.7                 1.5   
4          2.1             1.4                      18.8                 1.2   

   Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
0                             5.5              55.50              6.36   
1                             5.2              27.00              9.32   
2                             5.2              31.50             11.93   
3                             6.0              39.00             11.70   
4                             6.6              46.25              8.75   

  Certification Number Date Approved  
0                23-20    2023-01-23  
1                23-21    2023-01-23  
2                23-18    2023-01-23  
3                23-19    2023-01-23  
4                23-12    2023-01-16

## Compare Brand Names
We want to make sure that brand names match exactly for both dataframes.
1. create lists of brands for each dataframe
2. find the brands that are in both.
3. check if there are matches for that brand that are formatted differently
4. change the names of those brands and add them to the list

In [77]:
#1
#dimension brands
dim_brands = dimensions_df['Brand'].unique()
#flight num brands
fn_brands = flight_number_df['Brand'].unique()


In [78]:
#2
#find brands in both lists
same_brands = list(set(fn_brands).intersection(dim_brands))
print(f"There are now {len(same_brands)} brands")

There are now 38 brands


In [79]:
#3 
#compare lists with remaining flight number brands
remaining = np.setdiff1d(fn_brands,same_brands)
remaining

array(['Crosslap Discgolf Parks', 'Disc Golf UK', 'Guru Disc Golf',
       'Hyzer Bomb', 'Innova-Champion Discs', 'Millennium',
       'Obsidian Discs', 'Prodigy', 'Streamline', 'TOBU Discs',
       'UB Disc Golf', 'Viking Discs', 'Yikun Sports'], dtype=object)

In [80]:
#look at the dim_brands and see if any of thes match up, not sure if there is way to do this automatically
sorted(dim_brands)

['1080 Disc Golf',
 '3C Discs',
 'ABC Discs',
 'Above Ground Level',
 'Aerobie',
 'Albatross Disc Golf',
 'Alfa Discs',
 'All 4 Frisbee',
 'AquaFlight Discs',
 'Arsenal Discs',
 'Axiom Discs',
 'Best Disc Golf Discs',
 'Birdie',
 'Black Zombie Disc Golf',
 'CHING Sports',
 'Cheap Discs (Refuture Ventures Ab)',
 'Cheengz/Hyperflite',
 'Chumba Discs',
 'Clash Discs',
 'Crosslap Disc Golf Parks',
 'Crown Me (Sunny Sporting Co. Ltd)',
 'DKG Disc Sports',
 'DMI Sports',
 'Daredevil Discs',
 'Deity Discs',
 'Destiny/Dynamic Disc',
 'Disc Golf Association',
 'Disc Golf UK Ltd.',
 'Disc King',
 'Discmania',
 'Discraft',
 'Disctroyer OÜ',
 'Discwing, Ltd.',
 'Divergent Discs',
 'Doomsday Discs',
 'Dynamic Discs',
 'EMSCO Group',
 'EV-7',
 'Element Discs',
 'Elevation Disc Golf',
 'Essential Discs',
 'Eurodisc',
 'Exel Discs',
 'Ferris State University',
 'Finish Line Discs',
 'Fly High Discs',
 'FlyTec',
 'Fourth Circle Discs',
 'Franklin Sports',
 'Full Turn Discs',
 'GT Discs',
 'Galaxy Disc 

#### Alternate Names to be changed. this will make an easy dictionary.

'Crosslap Disc Golf Parks': 'Crosslap Discgolf Parks',
'Disc Golf UK Ltd.': 'Disc Golf UK',
'Guru':'Guru Disc Golf',
'Innova Champion Discs': 'Innova-Champion Discs',
'Innova Factory Store': 'Innova-Champion Discs',
'Millennium Golf Discs': 'Millennium',
'Obsidian Discs Oy': 'Obsidian Discs',
'Prodigy Disc': 'Prodigy'
'Streamline Discs': 'Streamline'
'Tobu Discs': 'TOBU Discs'
'UB Disc Golf - Hand Candy': 'UB Disc Golf',
'Viking Discs (IP-Agency Finland Oy)': 'Viking Discs'
'Yikun Discs':'Yikun Sports'

In [81]:
#4
#create dictionary from list above
brand_change_dict = {'Crosslap Disc Golf Parks': 'Crosslap Discgolf Parks', 'Disc Golf UK Ltd.': 'Disc Golf UK', 
                     'Guru':'Guru Disc Golf', 'Innova Champion Discs': 'Innova-Champion Discs', 
                     'Innova Factory Store': 'Innova-Champion Discs', 'Millennium Golf Discs': 'Millennium', 
                     'Obsidian Discs Oy': 'Obsidian Discs', 'Prodigy Disc': 'Prodigy', 'Streamline Discs': 'Streamline',
                     'Tobu Discs': 'TOBU Discs', 'UB Disc Golf - Hand Candy': 'UB Disc Golf', 
                     'Viking Discs (IP-Agency Finland Oy)': 'Viking Discs', 'Yikun Discs':'Yikun Sports'}
#change the dimensions_df to match flight_number_df
dimensions_df.replace({"Brand":brand_change_dict},inplace=True)
#test out the changes 
dimensions_df[dimensions_df['Brand'] == 'Innova-Champion Discs']

Brand            Mold  Max Weight (gr)  Diameter (cm)  \
31    Innova-Champion Discs           Alien            178.5           21.5   
62    Innova-Champion Discs       Rat (New)            176.0           21.2   
63    Innova-Champion Discs       Jay (New)            180.1           21.7   
64    Innova-Champion Discs  Wombat 3 (New)            180.1           21.7   
96    Innova-Champion Discs         Charger            175.1           21.1   
...                     ...             ...              ...            ...   
1623  Innova-Champion Discs           Coupe            176.0           21.2   
1629  Innova-Champion Discs           Aviar            176.0           21.2   
1632  Innova-Champion Discs        Aviar XD            176.0           21.2   
1633  Innova-Champion Discs            Aero            180.1           21.7   
1634  Innova-Champion Discs     Eagle (old)            180.1           21.7   

      Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
31            2.5             1.5                      20.3   
62            1.8             1.3                      18.8   
63            1.7             1.2                      18.9   
64            1.8             1.2                      18.9   
96            1.6             1.2                      16.4   
...           ...             ...                       ...   
1623          2.0             1.3                      19.4   
1629          2.0             1.5                      19.4   
1632          2.1             1.1                      19.4   
1633          2.5             1.3                      19.6   
1634          2.5             1.3                      19.6   

      Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
31                   0.6                             7.0              82.00   
62                   1.2                             6.1              66.75   
63                   1.4                             5.5              45.75   
64                   1.4                             5.5              30.00   
96                   2.3                             5.7              28.50   
...                  ...                             ...                ...   
1623                 0.9                             6.1              50.00   
1629                 0.9                             7.1              55.75   
1632                 0.9                             5.2              52.50   
1633                 1.0                             6.0              43.75   
1634                 1.0                             6.0              43.75   

      Flexibility (kg) Certification Number Date Approved  
31                3.30               22-228    2022-12-12  
62               11.14               22-188    2022-10-31  
63               11.48               22-187    2022-10-31  
64               11.59               22-189    2022-10-31  
96                8.98               22-148    2022-09-05  
...                ...                  ...           ...  
1623              8.16                85-01    1985-01-01  
1629             11.34                84-01    1984-01-01  
1632              9.10                84-08    1984-01-01  
1633              8.16                83-03    1983-06-01  
1634              8.16                83-05    1983-04-01  

[170 rows x 13 columns]

In [82]:
updated_dim_brands = dimensions_df['Brand'].unique()
add_brands = list(set(fn_brands).intersection(updated_dim_brands))
print(f"There are now {len(add_brands)} brands")

There are now 47 brands


## Fixing Mold Names
This will take a little more time. We will attempt to find some patterns and only worry about the discs in the dimensions_df that are in our final brand list.

In [83]:
ct = dimensions_df.shape[0]
print(f"There are currently {ct} discs in the dataframe")

There are currently 1674 discs in the dataframe


In [84]:
dim_df = dimensions_df[dimensions_df['Brand'].isin(add_brands)]
ct = dim_df.shape[0]
print(f"There are now {ct} discs in the dataframe")

There are now 1225 discs in the dataframe


## Fixing Mold Names

1. Eliminate discs in flight_number_df that are not in our same_brands list
2. Use difflib to find closest matches and then drop those don't match if it is not a huge number.

In [85]:
#1
#only include the brands in our list.
fn_df = flight_number_df[flight_number_df['Brand'].isin(add_brands)].copy()
dm_df = dimensions_df.copy()
fn_df.shape


(681, 9)

In [86]:
##problem disc as identified from the website change to radar
fn_df[fn_df['Mold']=='https://infinitediscs.com/Storm-Disc-Golf-Radar']

Brand                                             Mold    Type  \
740  Storm Disc Golf  https://infinitediscs.com/Storm-Disc-Golf-Radar  Putter   

    Speed Glide Turn Fade In Prod. Date Approved  
740   2.0   5.0  0.0  0.5      Yes    2019-09-05

In [87]:
fn_df.at[740,'Mold'] = 'Radar'
fn_df[fn_df['Mold']=='Radar']

Brand   Mold    Type Speed Glide Turn Fade In Prod.  \
740  Storm Disc Golf  Radar  Putter   2.0   5.0  0.0  0.5      Yes   

    Date Approved  
740    2019-09-05

## difflib! 
difflib can be used to match items from one column with the best match from another column.  

In [88]:
#2
#import difflib
import difflib
#add a duplicate column to keep track of original name
fn_df['Original_Mold'] = fn_df['Mold']
#convert the Brand in the fn_df dataset to closest match
fn_df['Mold'] = fn_df['Mold'].apply(lambda x: (difflib.get_close_matches(x, dm_df['Mold'])[:1] or [None])[0])
fn_df.head()


Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved    Original_Mold  
0    2012-09-12         Bee Line  
1    2010-05-29  Flying Squirrel  
2    2010-07-27          Mission  
3    2010-04-09            Money  
4    2010-05-29    Secret Weapon

In [89]:
#check the how many don't match with .shape
number = fn_df[fn_df['Mold'] != fn_df['Original_Mold']].shape[0]
print(f"There are {number} discs that do not match")

There are 56 discs that do not match


In [90]:
#drop these discs for now.  
filtered_fn_df = fn_df[fn_df['Original_Mold'] == fn_df['Mold']]
new_number = filtered_fn_df.shape[0]
print(f"There are {new_number} discs that were matched correctly")

There are 625 discs that were matched correctly


In [91]:
filtered_fn_df

Brand             Mold      Type Speed Glide  Turn Fade  \
0                ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0   
1                ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0   
2                ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0   
3                ABC Discs            Money    Putter   2.0   3.0   0.0  2.0   
4                ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0   
..                     ...              ...       ...   ...   ...   ...  ...   
737         Infinite Discs           Cohort    Putter   3.5   4.0   0.0  1.0   
738         Infinite Discs        Centurion   Fairway   7.0   5.0  -1.0  1.5   
739  Disc Golf Association             Rift  Midrange   5.0   4.0  -1.0  1.5   
740        Storm Disc Golf            Radar    Putter   2.0   5.0   0.0  0.5   
742             Mint Discs         Freetail  Distance  10.0   5.0  -3.0  2.0   

        In Prod. Date Approved    Original_Mold  
0            Yes    2012-09-12         Bee Line  
1            Yes    2010-05-29  Flying Squirrel  
2            Yes    2010-07-27          Mission  
3            Yes    2010-04-09            Money  
4            Yes    2010-05-29    Secret Weapon  
..           ...           ...              ...  
737          Yes    2019-07-10           Cohort  
738          Yes    2019-07-10        Centurion  
739          Yes    2019-08-01             Rift  
740          Yes    2019-09-05            Radar  
742  Coming Soon    2019-09-16         Freetail  

[625 rows x 10 columns]

## Fixing the ones that did not work.
These are the 40 discs that did not match.  This is managable to find a solution.

Below we will look at the molds that did not match and see if we can find a solution.  We will need to find disc mold names in the dim_df that match the Original_Mold column below.

In [92]:
unmatched_df = fn_df[fn_df['Original_Mold'] != fn_df['Mold']].copy()
unmatched_df


Brand                      Mold       Type Speed Glide  \
92               Discmania                      None   Distance  10.0   5.0   
93               Discmania                        D2    Fairway   9.0   5.0   
94               Discmania                        D3  Distance   11.0   5.0   
95               Discmania                      None   Distance  11.0   5.0   
96               Discmania                      None   Distance  13.0   5.0   
97               Discmania                       DD1   Distance  12.0   6.0   
98               Discmania                 Jackalope    Fairway   7.0   6.0   
99               Discmania                        F2    Fairway   7.0   4.0   
100              Discmania       D6 V2 (retooled D6)    Fairway   7.0   4.0   
101              Discmania                        F3    Fairway   9.0   4.0   
102              Discmania                      None   Midrange   4.0   5.0   
103              Discmania     MD1 (Midrange Driver)   Midrange   4.0   4.0   
104              Discmania                      None   Midrange   4.0   5.0   
105              Discmania     MD3 (Midrange Driver)   Midrange   5.0   5.0   
106              Discmania                        M4   Midrange   5.0   4.0   
107              Discmania                        M5   Midrange   5.0   4.0   
108              Discmania                     Panic     Putter   2.0   3.0   
109              Discmania                       PA1     Putter   2.0   3.0   
110              Discmania                      None     Putter   2.0   3.0   
111              Discmania      P3 (Putt & Approach)     Putter   3.0   2.0   
112              Discmania                       PD3     Putter   3.0   2.0   
113              Discmania                      None   Distance  10.0   4.0   
114              Discmania                       PD3   Distance  11.0   4.0   
115              Discmania                      None   Distance  12.0   4.0   
116              Discmania                      None   Distance  10.0   5.0   
117              Discmania                      None   Distance  10.0   5.0   
118              Discmania                      None    Fairway   9.0   5.0   
146               Discraft                    Knight   Distance  12.0   5.0   
367  Innova-Champion Discs                      None    Fairway   7.0   5.0   
628             TOBU Discs            Power Celerity    Fairway   8.0   4.0   
661    Westside Golf Discs  Queen (Pohjolan Emäntä1)   Distance  14.0   5.0   
663    Westside Golf Discs         Seer (Ennustaja1)    Fairway   7.0   5.0   
682           Yikun Sports                     Clash     Putter   1.0   2.0   
684           Yikun Sports                 Da'e (大鹗)   Distance  13.0   5.0   
685           Yikun Sports                      None    Fairway   8.0   5.0   
686           Yikun Sports                     Guide     Putter   2.0   3.0   
688           Yikun Sports                      Havu    Fairway   9.0   5.0   
689           Yikun Sports                  Jiao (狡)    Fairway   6.0   5.0   
690           Yikun Sports                      None   Distance  11.0   5.0   
691           Yikun Sports                       Koi   Midrange   5.0   5.0   
692           Yikun Sports                      Luna   Distance  14.0   5.0   
693           Yikun Sports         Twin Swords (双刃剑)   Distance  12.0   5.0   
694           Yikun Sports                      Vibe    Fairway   6.0   4.0   
695           Yikun Sports                      None   Distance  10.0   5.0   
696           Yikun Sports                 Wings (翼)   Midrange   3.0   3.0   
697           Yikun Sports                      None   Midrange   4.0   4.0   
712              Discmania                   Serpent   Distance  12.0   5.0   
713              Discmania                    Spirit   Midrange   6.0   4.0   
714              Discmania                  Invictus    Fairway   8.0   5.0   
715              Discmania                      None    Fairway   8.0  

In [93]:
#how many didn't match
print(f"{len(unmatched_df)} discs still need to be matched")

56 discs still need to be matched


## Remaining Mold Names

using the 'Original Mold', 'Brand', 'Date Approved' columns from the unmatched_df return all of the discs from the dimensions data that match the 'Brand' and 'Date Approved' columns then be able to select a matching disc if available.

In [94]:
test_dict = {0:'A',1:'B'}
## adding the none of the above option to a dictionary
def none_of(dictionary):
    new_key = list(dictionary)[-1] + 1
    dictionary[new_key] = 'None Of The Above'
    return dictionary
none_of(test_dict)

{0: 'A', 1: 'B', 2: 'None Of The Above'}

In [95]:
##function to return matching discs based on 'Brand', 'Date Approved' given a mold name
def find_molds(mold_name):
    #get the brand and the date
    disc_brand = unmatched_df.loc[unmatched_df['Original_Mold']==mold_name, 'Brand'].iat[0]
    disc_date = unmatched_df.loc[unmatched_df['Original_Mold']==mold_name, 'Date Approved'].array[0]
   
    #use brand and date to find all discs that match
    discs_df = dm_df[(dm_df['Brand']==disc_brand) & (dm_df['Date Approved']==disc_date)].reset_index(drop=True)
    disc_dict = dict(zip(discs_df.index, discs_df['Mold']))
    #add the "None of the Above" option and return
    choices =none_of(disc_dict)
    return choices
#test with a disc from unmatched_df above.
find_molds('Spring Ox')

{0: 'Magician (originally Fox Spirit)',
 1: 'Guardian Lion',
 2: 'Mentor (originally Sea Serpent)',
 3: 'Maestro (originally Spring Ox)',
 4: 'Genius (originally Sun Crow)',
 5: 'None Of The Above'}

## Picking the correct name from the list
Selecting the name from the dictionary and update the dm_df['Mold'] 

In [96]:
##define a function to select from a key dictionary and return the value
def dict_select(dictionary):
    print(dictionary)
    key = int(input("Pick the correct disc mold name: "))
    return dictionary[key]
dict_select(test_dict)

{0: 'A', 1: 'B', 2: 'None Of The Above'}
Pick the correct disc mold name: 1


'B'

## Putting it All Together

Now we can create a function to take the unmatched discs and change the 'Mold' column in the dm_df dataframe by using our find_molds function to match our 'Original_Mold' column.

In [97]:
unmatched_df.head()

Brand  Mold       Type Speed Glide  Turn Fade In Prod. Date Approved  \
92  Discmania  None   Distance  10.0   5.0  -1.0  2.0      Yes    2007-03-08   
93  Discmania    D2    Fairway   9.0   5.0  -1.0  2.0      Yes    2014-09-05   
94  Discmania    D3  Distance   11.0   5.0  -1.0  2.0      Yes    2017-03-08   
95  Discmania  None   Distance  11.0   5.0   0.0  2.0      Yes    2008-08-26   
96  Discmania  None   Distance  13.0   5.0  -2.0  2.0      Yes    2010-08-18   

    Original_Mold  
92     CD - Craze  
93            CD2  
94            CD3  
95  DD - Hysteria  
96     DD2 Frenzy

In [44]:
print(f"{len(unmatched_df)} unmatched discs")

56 unmatched discs


## Testing out the functionality of the Mold change

In [98]:
samples = ['Rei','Spring Ox', 'Sun Crow', 'Tomahawk']

for s in samples:
    try: 
        print(s)
        x = find_molds(s)
    except:
        x = {1:'Yes',2:'No'}
    print(dict_select(x))

Rei
{0: 'Rei (零)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0
Rei (零)
Spring Ox
{0: 'Magician (originally Fox Spirit)', 1: 'Guardian Lion', 2: 'Mentor (originally Sea Serpent)', 3: 'Maestro (originally Spring Ox)', 4: 'Genius (originally Sun Crow)', 5: 'None Of The Above'}
Pick the correct disc mold name: 3
Maestro (originally Spring Ox)
Sun Crow
{0: 'Magician (originally Fox Spirit)', 1: 'Guardian Lion', 2: 'Mentor (originally Sea Serpent)', 3: 'Maestro (originally Spring Ox)', 4: 'Genius (originally Sun Crow)', 5: 'None Of The Above'}
Pick the correct disc mold name: 4
Genius (originally Sun Crow)
Tomahawk
{0: 'Meteor Hammer (流星锤)', 1: 'Tomahawk (八卦钺)', 2: 'None Of The Above'}
Pick the correct disc mold name: 1
Tomahawk (八卦钺)


## Function to Replace Mismatched Names

This function will go through each disc in the unmatched_df and provide options using the 'Brand' and 'Date Approved'and the functions we created above.  

In [99]:
#updating the names in the dm_df with lambda function applied to our unmatched_df
def replace_mismatched(x):
    try:
        print(x)
        print("")
        dict2 = find_molds(x)
    except: 
        dict2 = {1:'Yes',2:'No'}
    response = dict_select(dict2)
    print("")
    return response

##apply to unmatched_df
unmatched_df['Mold_Change'] = unmatched_df['Original_Mold'].apply(lambda x: replace_mismatched(x))


CD - Craze

{0: 'CD (Craze, Control Driver)', 1: 'P1 (Maniac, Putt & Approach)', 2: 'P2 (Psycho, Pro Putter)', 3: 'None Of The Above'}
Pick the correct disc mold name: 0

CD2

{0: 'CD2 (Control Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

CD3

{0: 'CD3 (Control Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

DD - Hysteria

{0: 'DD (Hysteria, Distance Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

DD2 Frenzy

{0: 'DD2 (Frenzy, Distance Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

DDx

{0: 'DDx (Distance Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

FD - Jackal

{0: 'FD (Jackal, Fairway Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

FD2

{0: 'FD2 (Fairway Driver)', 1: 'None Of The Above'}
Pick the correct disc mold name: 0

FD2 (retooled)

{0: 'MD5 (Midrange Driver)', 1: 'FD2 (retooled Fairway Driver)', 2: 'None Of The Above'}
Pick the correct disc

## Make Mold Equal to Mold_Change and then merge the data.

change the Mold column in fn_df to Mold_Change in unmatched_df.  

In [100]:
unmatched_df.head()

Brand  Mold       Type Speed Glide  Turn Fade In Prod. Date Approved  \
92  Discmania  None   Distance  10.0   5.0  -1.0  2.0      Yes    2007-03-08   
93  Discmania    D2    Fairway   9.0   5.0  -1.0  2.0      Yes    2014-09-05   
94  Discmania    D3  Distance   11.0   5.0  -1.0  2.0      Yes    2017-03-08   
95  Discmania  None   Distance  11.0   5.0   0.0  2.0      Yes    2008-08-26   
96  Discmania  None   Distance  13.0   5.0  -2.0  2.0      Yes    2010-08-18   

    Original_Mold                     Mold_Change  
92     CD - Craze      CD (Craze, Control Driver)  
93            CD2            CD2 (Control Driver)  
94            CD3            CD3 (Control Driver)  
95  DD - Hysteria  DD (Hysteria, Distance Driver)  
96     DD2 Frenzy   DD2 (Frenzy, Distance Driver)

In [101]:
fn_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 681 entries, 0 to 742
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Brand          681 non-null    object        
 1   Mold           664 non-null    object        
 2   Type           681 non-null    object        
 3   Speed          681 non-null    object        
 4   Glide          681 non-null    object        
 5   Turn           681 non-null    object        
 6   Fade           681 non-null    object        
 7   In Prod.       681 non-null    object        
 8   Date Approved  677 non-null    datetime64[ns]
 9   Original_Mold  681 non-null    object        
dtypes: datetime64[ns](1), object(9)
memory usage: 78.5+ KB


In [102]:
## make a copy of fn_df so we don't have to keep rerunning the code to change molds 
flight_df = fn_df.copy()
flight_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved    Original_Mold  
0    2012-09-12         Bee Line  
1    2010-05-29  Flying Squirrel  
2    2010-07-27          Mission  
3    2010-04-09            Money  
4    2010-05-29    Secret Weapon

In [103]:
##create dictionary from unmatched key: original_mold, value: mold_change
mold_update_dict = pd.Series(unmatched_df.Mold_Change.values,index=unmatched_df.Original_Mold).to_dict()
mold_update_dict

{'CD - Craze': 'CD (Craze, Control Driver)',
 'CD2': 'CD2 (Control Driver)',
 'CD3': 'CD3 (Control Driver)',
 'DD - Hysteria': 'DD (Hysteria, Distance Driver)',
 'DD2 Frenzy': 'DD2 (Frenzy, Distance Driver)',
 'DDx': 'DDx (Distance Driver)',
 'FD - Jackal': 'FD (Jackal, Fairway Driver)',
 'FD2': 'FD2 (Fairway Driver)',
 'FD2 (retooled)': 'FD2 (retooled Fairway Driver)',
 'FD3': 'FD3 (Fairway Driver)',
 'MD': 'MD (Midrange Driver)',
 'MD1 - Midrange Driver': 'MD1 (Midrange Driver)',
 'MD2 - Fiend': 'MD2 (Fiend, Midrange Driver)',
 'MD3 - Midrange Driver': 'MD3 (Midrange Driver)',
 'MD4': 'MD4 (Midrange Driver)',
 'MD5': 'MD5 (Midrange Driver)',
 'P1 - Maniac': 'P1 (Maniac, Putt & Approach)',
 'P1x': 'P1x (Beaded Putter)',
 'P2 - Psycho': 'P2 (Psycho, Pro Putter)',
 'P3 - Putt & Approach': 'P3 (Putt & Approach)',
 'P3x': 'P3x (Putt & Approach)',
 'PD - Freak': 'PD (Freak, Power Driver)',
 'PDx': 'PDx (Power Driver)',
 'PD2 - Chaos': 'PD2 (Chaos, Power Driver)',
 'TD - Rush': 'TD (Rush, T

In [107]:
def mold_change_with_dict(x):
    if x in mold_update_dict:
        return mold_update_dict[x]
    return x
##change Mold using dictionary
flight_df['Mold'] = flight_df['Original_Mold'].apply(mold_change_with_dict)
flight_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved    Original_Mold  
0    2012-09-12         Bee Line  
1    2010-05-29  Flying Squirrel  
2    2010-07-27          Mission  
3    2010-04-09            Money  
4    2010-05-29    Secret Weapon

In [108]:
##check discmania because they had the most name issues
flight_df[flight_df['Brand']=='Discmania']

Brand                              Mold       Type Speed Glide  Turn  \
92   Discmania        CD (Craze, Control Driver)   Distance  10.0   5.0  -1.0   
93   Discmania              CD2 (Control Driver)    Fairway   9.0   5.0  -1.0   
94   Discmania              CD3 (Control Driver)  Distance   11.0   5.0  -1.0   
95   Discmania    DD (Hysteria, Distance Driver)   Distance  11.0   5.0   0.0   
96   Discmania     DD2 (Frenzy, Distance Driver)   Distance  13.0   5.0  -2.0   
97   Discmania             DDx (Distance Driver)   Distance  12.0   6.0  -1.0   
98   Discmania       FD (Jackal, Fairway Driver)    Fairway   7.0   6.0  -1.0   
99   Discmania              FD2 (Fairway Driver)    Fairway   7.0   4.0   0.0   
100  Discmania     FD2 (retooled Fairway Driver)    Fairway   7.0   4.0   0.0   
101  Discmania              FD3 (Fairway Driver)    Fairway   9.0   4.0   0.0   
102  Discmania              MD (Midrange Driver)   Midrange   4.0   5.0   0.0   
103  Discmania             MD1 (Midrange Driver)   Midrange   4.0   4.0   0.0   
104  Discmania      MD2 (Fiend, Midrange Driver)   Midrange   4.0   5.0   0.0   
105  Discmania             MD3 (Midrange Driver)   Midrange   5.0   5.0   0.0   
106  Discmania             MD4 (Midrange Driver)   Midrange   5.0   4.0   0.0   
107  Discmania             MD5 (Midrange Driver)   Midrange   5.0   4.0   0.0   
108  Discmania      P1 (Maniac, Putt & Approach)     Putter   2.0   3.0   0.0   
109  Discmania               P1x (Beaded Putter)     Putter   2.0   3.0   0.0   
110  Discmania           P2 (Psycho, Pro Putter)     Putter   2.0   3.0   0.0   
111  Discmania              P3 (Putt & Approach)     Putter   3.0   2.0   0.0   
112  Discmania             P3x (Putt & Approach)     Putter   3.0   2.0   0.0   
113  Discmania          PD (Freak, Power Driver)   Distance  10.0   4.0   0.0   
114  Discmania                PDx (Power Driver)   Distance  11.0   4.0   0.0   
115  Discmania         PD2 (Chaos, Power Driver)   Distance  12.0   4.0   0.0   
116  Discmania            TD (Rush, Turn Driver)   Distance  10.0   5.0  -2.0   
117  Discmania          TD2 (Fever, Turn Driver)   Distance  10.0   5.0  -4.0   
118  Discmania                 TDx (Turn Driver)    Fairway   9.0   5.0  -3.0   
712  Discmania   Mentor (originally Sea Serpent)   Distance  12.0   5.0  -2.0   
713  Discmania  Magician (originally Fox Spirit)   Midrange   6.0   4.0   0.0   
714  Discmania               Instinct (150-175g)    Fairway   8.0   5.0   0.0   
715  Discmania      Genius (originally Sun Crow)    Fairway   8.0   5.0  -4.0   
716  Discmania    Maestro (originally Spring Ox)   Midrange   4.0   3.0   0.0   
717  Discmania                     Guardian Lion     Putter   1.0   3.0   0.0   
718  Discmania                      DD3 (retest)   Distance  12.0   5.0  -1.0   
724  Discmania                            Method   Midrange   5.0   5.0   0.0   
728  Discmania                            Enigma   Distance  13.0   5.0  -1.0   
731  Discmania                              Link     Putter   2.0   3.0   0.0   

    Fade In Prod. Date Approved          Original_Mold  
92   2.0      Yes    2007-03-08             CD - Craze  
93   2.0      Yes    2014-09-05                    CD2  
94   2.0      Yes    2017-03-08                    CD3  
95   2.0      Yes    2008-08-26          DD - Hysteria  
96   2.0      Yes    2010-08-18             DD2 Frenzy  
97   2.0      Yes    2016-02-25                    DDx  
98   1.0      Yes    2011-11-13            FD - Jackal  
99   2.0      Yes    2014-03-24                    FD2  
100  2.0      Yes    2018-03-29         FD2 (retooled)  
101  3.0      Yes    2015-06-26                    FD3  
102  0.0      Yes    2015-04-29                     MD  
103  2.0      Yes    2006-11-14  MD1 - Midrange Driver  
104  2.0      Yes    2010-04-04            MD2 - Fiend  
105  2.0      Yes    2013-03-29  MD3 - Midrange Driver  
106  3.0      Yes    2017-04-21                    MD4  
107  2.0      Yes 

It looks like we are in business...

In [110]:
flight_df = flight_df.drop(['Original_Mold'],axis=1)
flight_df.head()

Brand             Mold      Type Speed Glide  Turn Fade In Prod.  \
0  ABC Discs         Bee Line  Distance  14.0   5.0  -3.0  2.0      Yes   
1  ABC Discs  Flying Squirrel  Midrange   6.0   5.0  -3.0  2.0      Yes   
2  ABC Discs          Mission  Midrange   4.0   5.0  -1.0  2.0      Yes   
3  ABC Discs            Money    Putter   2.0   3.0   0.0  2.0      Yes   
4  ABC Discs    Secret Weapon  Distance  10.0   5.0  -3.0  2.0      Yes   

  Date Approved  
0    2012-09-12  
1    2010-05-29  
2    2010-07-27  
3    2010-04-09  
4    2010-05-29

## Finally Merging The Data Frames
flight_df merged with dm_df

In [111]:
dm_df.head()

Brand              Mold  Max Weight (gr)  Diameter (cm)  \
0    Full Turn Discs             Guide            180.9           21.8   
1  Wing It Disc Golf            Dream             176.8           21.3   
2      Dynamic Discs   Sockibomb Felon            176.0           21.2   
3      Dynamic Discs  Supreme Fugitive            180.1           21.7   
4      STM Disc Golf      Cloud Ripper            175.1           21.1   

   Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  Rim Thickness (cm)  \
0          1.8             1.2                      19.0                 1.4   
1          1.4             1.1                      16.5                 2.4   
2          1.6             1.1                      17.1                 2.0   
3          1.9             1.3                      18.7                 1.5   
4          2.1             1.4                      18.8                 1.2   

   Rim Depth / Diameter Ratio (%)  Rim Configuration  Flexibility (kg)  \
0                             5.5              55.50              6.36   
1                             5.2              27.00              9.32   
2                             5.2              31.50             11.93   
3                             6.0              39.00             11.70   
4                             6.6              46.25              8.75   

  Certification Number Date Approved  
0                23-20    2023-01-23  
1                23-21    2023-01-23  
2                23-18    2023-01-23  
3                23-19    2023-01-23  
4                23-12    2023-01-16

In [113]:
merged_df = pd.merge(flight_df, dm_df, on=['Mold','Brand','Date Approved'])
merged_df

Brand                                 Mold      Type  \
0                ABC Discs                             Bee Line  Distance   
1                ABC Discs                      Flying Squirrel  Midrange   
2                ABC Discs                              Mission  Midrange   
3                ABC Discs                                Money    Putter   
4                ABC Discs                        Secret Weapon  Distance   
..                     ...                                  ...       ...   
664         Infinite Discs                            Centurion   Fairway   
665  Disc Golf Association                                 Rift  Midrange   
666        Storm Disc Golf                                Radar    Putter   
667               Discraft  Malta (Paul McBeth Proto Mid-Range)  Midrange   
668             Mint Discs                             Freetail  Distance   

    Speed Glide  Turn Fade     In Prod. Date Approved  Max Weight (gr)  \
0    14.0   5.0  -3.0  2.0          Yes    2012-09-12            175.1   
1     6.0   5.0  -3.0  2.0          Yes    2010-05-29            181.8   
2     4.0   5.0  -1.0  2.0          Yes    2010-07-27            180.9   
3     2.0   3.0   0.0  2.0          Yes    2010-04-09            174.3   
4    10.0   5.0  -3.0  2.0          Yes    2010-05-29            175.1   
..    ...   ...   ...  ...          ...           ...              ...   
664   7.0   5.0  -1.0  1.5          Yes    2019-07-10            175.1   
665   5.0   4.0  -1.0  1.5          Yes    2019-08-01            179.3   
666   2.0   5.0   0.0  0.5          Yes    2019-09-05            175.1   
667   5.0   4.0   1.0  3.0          Yes    2019-09-12            176.8   
668  10.0   5.0  -3.0  2.0  Coming Soon    2019-09-16            175.1   

     Diameter (cm)  Height (cm)  Rim Depth (cm)  Inside Rim Diameter (cm)  \
0             21.1          1.6             1.1                      16.2   
1             21.9          2.0             1.4                      19.4   
2             21.8          2.0             1.4                      19.4   
3             21.0          2.2             1.7                      19.0   
4             21.1          2.0             1.1                      17.3   
..             ...          ...             ...                       ...   
664           21.1          1.8             1.2                      17.8   
665           21.6          1.8             1.3                      18.6   
666           21.1          2.0             1.5                      19.1   
667           21.3          1.7             1.3                      18.6   
668           21.1          1.7             1.2                      17.1   

     Rim Thickness (cm)  Rim Depth / Diameter Ratio (%)  Rim Configuration  \
0                   2.5                             5.2              26.25   
1                   1.3                             6.4              30.75   
2                   1.2                             6.4              39.25   
3                   1.0                             8.1              56.50   
4                   1.9                             5.2              26.75   
..                  ...                             ...                ...   
664                 1.7                             5.7              37.00   
665                 1.5                             6.0              47.00   
666                 1.0                             7.1              55.50   
667                 1.3                             6.1              45.00   
668                 2.0                             5.7              37.75   

     Flexibility (kg) Certification Number  
0                8.39                12-36  
1                3.52                10-13  
2                6.12                10-22  
3               11.11                10-08  
4                7.26                10-14  
..                ...                  ...  
664              8.75         

In [114]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 669 entries, 0 to 668
Data columns (total 19 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   Brand                           669 non-null    object        
 1   Mold                            669 non-null    object        
 2   Type                            669 non-null    object        
 3   Speed                           669 non-null    object        
 4   Glide                           669 non-null    object        
 5   Turn                            669 non-null    object        
 6   Fade                            669 non-null    object        
 7   In Prod.                        669 non-null    object        
 8   Date Approved                   669 non-null    datetime64[ns]
 9   Max Weight (gr)                 669 non-null    float64       
 10  Diameter (cm)                   669 non-null    float64       
 11  Height

In [116]:
##save data to prevent having to go through those steps again
merged_df.to_csv('data/flight_and_dimension.csv', sep='\t', encoding='utf-8', index=False)

## Workbook Results
Merged the dataframes into a csv file for further analysis